analytics/analytics.go

```diff
@@ -0,0 +1,258 @@
+}
+
+// IsConsentRecorded returns if valid consent has been recorded for tracking
+func IsConsentRecorded() (rec bool) {
+	if len(consentPath) == 0 {
+		usr, err := user.Current()
+		if err != nil {
+			return false
+		}
+		consentPath = filepath.Join(usr.HomeDir, ".config", "section", "cli_tracking_consent")
+	}
+	if _, err := os.Stat(consentPath); err != nil {
+		return false
+	}
+
+	consentFile, err := os.Open(consentPath)
+	if err != nil {
+		return false
+	}
+	defer consentFile.Close()
+
+	contents, err := ioutil.ReadAll(consentFile)
+	if err != nil {
+		return false
+	}
+
+	var consent cliTrackingConsent
+	err = json.Unmarshal(contents, &consent)
+
+	return err == nil
+}
+
+// ReadConsent finds if consent has been given, either from file or by prompt.
+func ReadConsent() {
+	if !IsConsentRecorded() {
+		PromptForConsent()
+	}
+
+	if _, err := os.Stat(consentPath); err != nil {
+		return
+	}
+
+	consentFile, err := os.Open(consentPath)
+	if err != nil {
+		return
+	}
+	defer consentFile.Close()
+
+	var consent cliTrackingConsent
+	contents, err := ioutil.ReadAll(consentFile)
+	if err != nil {
+		return
+	}
+	err = json.Unmarshal(contents, &consent)
+	if err != nil {
+		return
+	}
+
+	ConsentGiven = consent.ConsentGiven
+}
+
+// Println formats using the default formats for its operands and writes to output.
+// Output can be overridden for testing purposes by setting: analytics.out
+// It returns the number of bytes written and any write error encountered.
+func Println(a ...interface{}) (n int, err error) {
+	return fmt.Fprintln(out, a...)
+}
+
+// Printf formats according to a format specifier and writes to standard output.
+// Output can be overridden for testing purposes by setting: analytics.out
+// It returns the number of bytes written and any write error encountered.
+func Printf(format string, a ...interface{}) (n int, err error) {
+	return fmt.Fprintf(out, format, a...)
+}
+
+// PromptForConsent interactively prompts the user for consent
+func PromptForConsent() {
+	Printf("👋 Hi!\n\n")
+	Printf("Thanks for using the Section CLI.\n\n")
+	Printf("We're still working out how to create the best experience for you.\n\n")
+	Printf("We'd like to collect some anonymous info about how you use the CLI.\n\n")
+	Printf("Are you OK with this? [y/N] ")
+
+	reader := bufio.NewReader(in)
+	text, err := reader.ReadString('\n')
+	if err != nil {
+		Println("Error: unable to read your response. Exiting.")
+		os.Exit(2)
+	}
+	text = strings.Replace(text, "\n", "", -1) // convert CRLF to LF
+
+	if strings.EqualFold(text, "y") {
+		Printf("\nThank you!\n")
+		ConsentGiven = true
+	} else {
+		ConsentGiven = false
+		Printf("\nNo worries! We won't ask again.\n")
+	}
+
+	WriteConsent()
+}
+
+// WriteConsent writes the current consent state to a persistent file
+func WriteConsent() {
+	c := cliTrackingConsent{ConsentGiven: ConsentGiven}
+	json, err := json.Marshal(c)
+	if err != nil {
+		Println("Error: unable to record consent. Exiting.")
```

Comment:
Both this error and the next one are the same. This doesn't help you much with debugging in case of an error.

Reply:
Good call, I refactored in ab7d68f1427e248da09fdfefc19523ca74eb5211 to handle the error in one place.